src/controller/python/chip-device-ctrl.py

```diff
@@ -395,11 +397,78 @@ def do_connect(self, line):
+        except ChipExceptions.ChipStackException as ex:
             print(str(ex))
             return
         print("Connected")
 
+    def do_zcl(self, line):
+        """
+        To send ZCL message to device:
+        zcl <cluster> <command> <nodeid> <endpoint> <groupid> [key=value]...
+        To get a list of clusters:
+        zcl ?
+        To get a list of commands in cluster:
+        zcl ? <cluster>
+
+        Send ZCL command to device nodeid
+        """
+        try:
+            args = shlex.split(line)
+            if len(args) == 1 and args[0] == '?':
+                print(self.devCtrl.ZCLList().keys())
+            elif len(args) == 2 and args[0] == '?':
+                cluster = self.devCtrl.ZCLList().get(args[1], None)
+                if not cluster:
+                    print("UnknownCluster: {}".format(cluster))
+                    return
+                for commands in cluster.items():
+                    args = ", ".join(["{}: {}".format(argName, argType)
+                                      for argName, argType in commands[1].items()])
+                    print(commands[0])
+                    if commands[1]:
+                        print("  ", args)
+                    else:
+                        print("  <no arguments>")
+            elif len(args) > 4:
+                commandArgs = {}
+                cluster = self.devCtrl.ZCLList().get(args[0], None)
+                if not cluster:
+                    raise ChipExceptions.UnknownCluster(args[0])
+                command = cluster.get(args[1], None)
+                if not cluster:
+                    raise ChipExceptions.UnknownCommand(args[0], args[1])
+                for kvPair in args[5:]:
+                    if kvPair.find("=") < 0:
+                        print("Argument should in key=value format")
+                        return
+                    key, value = kvPair.split("=", 1)
+                    valueType = command.get(key, None)
+                    if valueType == 'int':
+                        commandArgs[key] = int(value)
+                    elif valueType == 'str':
+                        commandArgs[key] = value
+                    elif valueType == 'bytes':
+                        enc, encValue = value.split(":", 1)
+                        if enc == "str":
+                            commandArgs[key] = encValue.encode(
+                                "utf-8") + b'\x00'
+                        elif enc == "hex":
+                            commandArgs[key] = bytes.fromhex(encValue)
+                        else:
+                            print(
+                                "Unsupported encoding, supported encoding: str, hex")
+                            return
+                self.devCtrl.ZCLSend(args[0], args[1], int(
+                    args[2]), int(args[3]), int(args[4]), commandArgs)
+            else:
+                self.do_help("zcl")
+                return
```

Comment:
return does not seem needed here nor in the except block below.

Nesting within this method makes it hard to read/follow for me. 

Could we split functionality up?  Like one method for 'first arg is ?' another one for 4+ args and dohelp otherwise.